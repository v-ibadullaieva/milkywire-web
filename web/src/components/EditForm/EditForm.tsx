import React, { useEffect, useState, FormEvent } from 'react';
import { Form, Input, Button, Upload, Icon, message } from 'antd';
import { WrappedFormUtils, FormComponentProps } from 'antd/lib/form/Form';

import { uploadImg } from '../../api/posts';
import { PostType } from '../PostsList/PostsList';

interface EditFormProps extends FormComponentProps<any> {
  post: PostType,
  form: WrappedFormUtils,
  onCancel: () => void,
  editPost: (post: PostType) => void
}

export interface UpdatedPostType {
  description: string,
  data: {
    media: [{
      image: string,
      width: number,
      height: number,
      version: string,
      description: string
    }]
  },
}

type valuesType = {
  description: string,
  mediaDescription: string,
  height: number,
  width: number,
  version: string,
}

const hasErrors = (fieldsError: any) => {
  return Object.keys(fieldsError).some(field => fieldsError[field]);
}

const beforeUpload = (file: any) => {
  const isJpgOrPng = file.type === 'image/jpeg' || file.type === 'image/png';
  if (!isJpgOrPng) {
    message.error('You can only upload JPG/PNG file!');
  }
  const isLt2M = file.size / 1024 / 1024 < 2;
  if (!isLt2M) {
    message.error('Image must smaller than 2MB!');
  }
  return isJpgOrPng && isLt2M;
}

const EditForm = (props: EditFormProps) => {
  const [loading] = useState(false);
  const [postThumb, setPostThumb] = useState(props.post.data.media[0].image || null);

  useEffect(() => {
    props.form.validateFields();
    return () => props.form.resetFields();
  }, [props.post.id]);

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    props.form.validateFields((err: Error, values: valuesType) => {
      if (!err) {
        const formatValues: UpdatedPostType = {
          description: values.description,
          data: {
            media: [{
              ...props.post.data.media[0],
              description: values.mediaDescription,
              height: values.height,
              width: values.width,
              version: values.version,
            }]
          }
        }

        props.editPost({ ...props.post, ...formatValues });

        props.form.resetFields();
        props.onCancel();
      }
    });
  };

  const handleImgUpload = async ({ file, onError }: any) => {
    try {
      const imgLink = await uploadImg(file);
      const data = {
        data: {
          media: [{
            ...props.post.data.media[0],
            image: imgLink.url
          }]
        }
      };

      setPostThumb(imgLink.url);
      props.editPost({ ...props.post, ...data } as PostType);

      if (!imgLink) {
        message.error('Something went wrong, contact your administrator or try again');
        return;
      }

      message.success(`${file.name} file uploaded successfully.`);
    } catch (e) {
      onError(e);
    }
  };

  const uploadButton = (
    <div>
      <Icon type={loading ? 'loading' : 'plus'} />
      <div className='ant-upload-text'>Upload</div>
    </div>
  );

  const { getFieldDecorator, getFieldsError, isFieldsTouched } = props.form;
  const { data: { media }, description } = props.post;

  return (
    <Form onSubmit={handleSubmit}>
      <Upload
        name='img'
        listType='picture-card'
        showUploadList={false}
        beforeUpload={beforeUpload}
        customRequest={file => handleImgUpload(file)}
      >
        {postThumb ? <img src={postThumb} alt='avatar' style={{ width: '100%' }} /> : uploadButton}
      </Upload>
      <Form.Item label='Description'>
        {getFieldDecorator('description', {
          initialValue: description ? description : 'Please enter description',
          rules: [{ required: true, message: 'Please enter description' }],
        })(
          <Input.TextArea placeholder='Enter description' />
        )}
      </Form.Item>
      <Form.Item label='Media description'>
        {getFieldDecorator('mediaDescription', {
          initialValue: media[0].description ? media[0].description : 'Please enter media description',
          rules: [{ required: true, message: 'Please enter media description' }],
        })(
          <Input.TextArea placeholder='Enter media description' />
        )}
      </Form.Item>
      <Form.Item label='Media width'>
        {getFieldDecorator('width', {
          initialValue: media[0].width ? media[0].width : 'Please enter media width',
          rules: [{ required: true, message: 'Please enter media width' }],
        })(
          <Input
            type='width'
            placeholder='Enter media width'
          />
        )}
      </Form.Item>
      <Form.Item label='Media height'>
        {getFieldDecorator('height', {
          initialValue: media[0].height ? media[0].height : 'Please enter media height',
          rules: [{ required: true, message: 'Please enter media height' }],
        })(
          <Input
            type='height'
            placeholder='Enter media height'
          />
        )}
      </Form.Item>
      <Form.Item label='Media version'>
        {getFieldDecorator('version', {
          initialValue: media[0].version ? media[0].version : 'Please enter media versiont',
          rules: [{ required: true, message: 'Please enter media version' }],
        })(
          <Input
            type='version'
            placeholder='Enter media version'
          />
        )}
      </Form.Item>
      <Form.Item>
        <Button
          type='primary'
          htmlType='submit'
          disabled={!isFieldsTouched() || hasErrors(getFieldsError())}
        >
          Update
        </Button>
      </Form.Item>
    </Form>
  )
}

const WrappedForm = Form.create<EditFormProps>({ name: 'EditForm' })(EditForm);

export default WrappedForm;