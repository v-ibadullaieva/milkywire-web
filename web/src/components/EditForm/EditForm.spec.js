import React from 'react';
import { create } from "react-test-renderer";

import EditForm from './EditForm';

const POST = {
  id: 0,
  type: "IMAGES",
  description: "lol",
  data: {
    media: [{
      image: "https://picsum.photos/id/0/5616/3744",
      width: 5616,
      height: 3744,
      version: "2019-03-14",
      description: "Alejandro Escamilla"
    }]
  },
  impacter_id: 0
};

describe('EditForm component', () => {
  test("Matches the snapshot", () => {
    const post = create(<EditForm post={POST} />);
    expect(post.toJSON()).toMatchSnapshot();
  });
});
