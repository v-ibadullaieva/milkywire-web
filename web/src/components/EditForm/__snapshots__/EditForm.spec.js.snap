// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm component Matches the snapshot 1`] = `
<form
  className="ant-form ant-form-horizontal"
  onSubmit={[Function]}
>
  <span
    className="ant-upload-picture-card-wrapper"
  >
    <div
      className="ant-upload ant-upload-select ant-upload-select-picture-card"
    >
      <span
        className="ant-upload"
        onClick={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex="0"
      >
        <input
          accept=""
          directory={null}
          multiple={false}
          onChange={[Function]}
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="file"
          webkitdirectory={null}
        />
        <img
          alt="avatar"
          src="https://picsum.photos/id/0/5616/3744"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </span>
    </div>
  </span>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="EditForm_description"
        onClick={[Function]}
        title="Description"
      >
        Description
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
      style={Object {}}
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <textarea
            className="ant-input"
            data-__field={
              Object {
                "dirty": false,
                "errors": undefined,
                "name": "description",
                "validating": false,
                "value": "lol",
              }
            }
            data-__meta={
              Object {
                "initialValue": "lol",
                "name": "description",
                "originalProps": Object {
                  "placeholder": "Enter description",
                  "type": "description",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please enter description",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please enter description",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="EditForm_description"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter description"
            style={Object {}}
            type="description"
            value="lol"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="EditForm_mediaDescription"
        onClick={[Function]}
        title="Media description"
      >
        Media description
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
      style={Object {}}
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <textarea
            className="ant-input"
            data-__field={
              Object {
                "dirty": false,
                "errors": undefined,
                "name": "mediaDescription",
                "validating": false,
                "value": "Alejandro Escamilla",
              }
            }
            data-__meta={
              Object {
                "initialValue": "Alejandro Escamilla",
                "name": "mediaDescription",
                "originalProps": Object {
                  "placeholder": "Enter media description",
                  "type": "mediaDescription",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please enter media description",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please enter media description",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="EditForm_mediaDescription"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter media description"
            style={Object {}}
            type="mediaDescription"
            value="Alejandro Escamilla"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="EditForm_width"
        onClick={[Function]}
        title="Media width"
      >
        Media width
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
      style={Object {}}
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "dirty": false,
                "errors": undefined,
                "name": "width",
                "validating": false,
                "value": 5616,
              }
            }
            data-__meta={
              Object {
                "initialValue": 5616,
                "name": "width",
                "originalProps": Object {
                  "placeholder": "Enter media width",
                  "type": "width",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please enter media width",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please enter media width",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="EditForm_width"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter media width"
            type="width"
            value={5616}
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="EditForm_height"
        onClick={[Function]}
        title="Media height"
      >
        Media height
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
      style={Object {}}
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "dirty": false,
                "errors": undefined,
                "name": "height",
                "validating": false,
                "value": 3744,
              }
            }
            data-__meta={
              Object {
                "initialValue": 3744,
                "name": "height",
                "originalProps": Object {
                  "placeholder": "Enter media height",
                  "type": "height",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please enter media height",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please enter media height",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="EditForm_height"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter media height"
            type="height"
            value={3744}
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="EditForm_version"
        onClick={[Function]}
        title="Media version"
      >
        Media version
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
      style={Object {}}
    >
      <div
        className="ant-form-item-control has-success"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "dirty": false,
                "errors": undefined,
                "name": "version",
                "validating": false,
                "value": "2019-03-14",
              }
            }
            data-__meta={
              Object {
                "initialValue": "2019-03-14",
                "name": "version",
                "originalProps": Object {
                  "placeholder": "Enter media version",
                  "type": "version",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please enter media version",
                    "required": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please enter media version",
                        "required": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="EditForm_version"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter media version"
            type="version"
            value="2019-03-14"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-control-wrapper"
      style={Object {}}
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <button
            className="ant-btn ant-btn-primary"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            <span>
              Update
            </span>
          </button>
        </span>
      </div>
    </div>
  </div>
</form>
`;
